### PR TITLE
Take loan feature

### LINKED ISSUE #671: Loan: take loan

### PR DESCRIPTION
**Pull request by Mixa84**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:
Initial development on `takeloan` function to "mint" tokens as a loan based on deposited collateral in vault.

#### Which issue(s) does this PR fixes?:

Fixes #671

### CODE REVIEW

**Comment by Mixa84** [comment]
> So the difference is it may not contain loan tokens ? I don't see reason this function to exist, it's used in `getloan` if there is no loan tokens who cares for collaterals ?

The point was that `CalculateCollateralizationRatio` was not returning anything if there were no loans taken. I wanted to display total collateral value and total loan value. If we don't really need to see how much is in collateral if there is no loan then we can remove this function and I'll use `CalculateCollateralizationRatio`.

**Review by dcorral — APPROVED** [review]

File: src/masternodes/rpc_loan.cpp [hunk 1]
```diff
@@ -970,6 +978,157 @@ UniValue getloanscheme(const JSONRPCRequest& request) {
...
+        auto collateral = pcustomcsview->GetCollateralAndLoanValue(vaultId, collaterals, height);
+
+        totalCollaterals += collateral->totalCollaterals();
+        totalLoans += collateral->totalLoans();
+
+        return true;
+    });
+
+    ret.pushKV("Collateral value (USD)",ValueFromAmount(totalCollaterals));
+    ret.pushKV("Loan value (USD)",ValueFromAmount(totalLoans));
```
Comment (ShengguangXiao) [inline src/masternodes/rpc_loan.cpp:1126]: Seems we don't use blanks in our keys.

Comment (Mixa84) [reply src/masternodes/rpc_loan.cpp:1126]: Forgot, I will correct it.

**Comment by bvbfan** [comment]
> > So the difference is it may not contain loan tokens ? I don't see reason this function to exist, it's used in `getloan` if there is no loan tokens who cares for collaterals ?
> 
> The point was that `CalculateCollateralizationRatio` was not returning anything if there were no loans taken. I wanted to display total collateral value and total loan value. If we don't really need to see how much is in collateral if there is no loan then we can remove this function and I'll use `CalculateCollateralizationRatio`.

Yep, there is nothing to calculate when there is no loan tokens, it should be the same for `getloan`, there is nothing to return when no loans.

File: src/masternodes/masternodes.cpp [hunk 2]
```diff
@@ -847,6 +851,45 @@ boost::optional<CCollateralLoans> CCustomCSView::CalculateCollateralizationRatio
...
+            assert(rate && rate->height <= height);
+            auto oracle = GetOracleData(token->priceFeedTxid);
+            assert(oracle);
+            auto price = GetOraclePriceUSD(*oracle.val, token->symbol);
+            auto value = loan.second + rate->interestToHeight + ((height - rate->height + 1) * rate->interestPerBlock);
+            ret.loans.push_back({loan.first, MultiplyAmounts(price, value)});
+        }
+    }
+
     return ret;
```
Comment (bvbfan) [inline src/masternodes/masternodes.cpp:893]: Merge this function with main one.

File: src/masternodes/masternodes.h [hunk 3]
```diff
@@ -287,6 +287,7 @@ struct CCollateralLoans {
     std::vector<CTokenAmount> loans; // in USD
 
     uint32_t ratio() const {
+        if (loans.empty()) return 0;
```
Comment (bvbfan) [inline src/masternodes/masternodes.h:290]: Return `std::numeric_limits<uint32_t>::max` because when only collaterals are in ratio is the higher one.

**Review by bvbfan — CHANGES_REQUESTED** [review]

**Comment by bvbfan** [comment]
Add `LoanTakeLoanCreationTx` and `LoanTakeLoanVaultKey` to `CheckPrefix` `CLoanView`

**Review by bvbfan — APPROVED** [review]

**Pull request merged**